Fix insert step and allow "closing" it

HarshKhandeparkar opened the pull request:
Fixes https://github.com/publiclab/image-sequencer/issues/640
Fixes #918
Fixes #919

Insert step button no longer appends more than once.


Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/reviewers` for help, in a comment below

Thanks!

Divy123 commented on examples/lib/intermediateHtmlStepUi.js:
```diff
@@ -60,11 +60,29 @@ function IntermediateHtmlStepUi(_sequencer, step, options) {
+        $(step.ui.querySelector('.insert-text')).fadeToggle(200, () => {$(step.ui.querySelector('.no-insert-text')).fadeToggle(200)})
+      }
+      else {
+        $(step.ui.querySelector('.no-insert-text')).fadeToggle(200, () => {$(step.ui.querySelector('.insert-text')).fadeToggle(200)})
```
I think we should migrate to normal functions here instead of arrow functions as it may not be supported by some browsers.

HarshKhandeparkar replied:
I had asked @jywarren about this, he said I can use these if they are supported and they are.

I think all modern (>IE) browsers support arrow functions and template literals. Including EdgeSafari everything.

https://kangax.github.io/compat-table/es6/

HarshKhandeparkar commented:
@Divy123 arrow functions are supported by all browsers excluding IE

Divy123 commented:
Still I think leaving an exception is not a good practice.

HarshKhandeparkar commented:
> Still I think leaving an exception is not a good practice.

IE is not used by anybody and even if somebody uses it I don't think ImageSequencer itself will work on it. Also, arrow functions have been used earlier in some PRs and they are in some `module.js` files and also in `defaultHtmlStepUi.js` file too I think. Or maybe in `demo.js`. I have seen them somewhere for sure.

Divy123 commented:
Can you please find out those snippets so that things can be rectified.
Also one of my friends has tested it on IE.

Divy123 commented:
@jywarren I think this idea is quite good. Maybe we can use Babel for this issue?

HarshKhandeparkar commented:
https://www.w3schools.com/browsers/ refers to Edge and IE as the same entity though they are not. Edge is updated and IE is not. Anyways whoever uses default windows10 browser uses Edge, therefore, it is the third most used. But if we are going to use babel then we don't need to worry. This can be merged and then compiled in a separate PR.

Divy123 commented:
Sure but as of now even I was directed to use normal functions and I don't think it to be a difficult task.

HarshKhandeparkar commented:
> Sure but as of now even I was directed to use normal functions on one of my PR and I don't think it to be a difficult task.

Ok I'll do that.

HarshKhandeparkar commented:
@publiclab/is-reviewers please review this.

HarshKhandeparkar commented:
@Divy123 I have made the changes. Please check them out.

HarshKhandeparkar commented:
@Divy123 can you open the issue for babel? We can discuss everything there. This might benefit us a lot. Will make the source code a lot cleaner and simpler.

HarshKhandeparkar commented:
@jywarren please check this out.

jywarren commented:
This looks great. Are we all good to merge here once it's been rebased? Thank you!!!

HarshKhandeparkar commented:
My exams are starting. So I will not be available until 30th January. I will fix the conflicts later.

HarshKhandeparkar commented:
I have merged with main. @publiclab/is-reviewers please review.

HarshKhandeparkar commented:
@publiclab/is-reviewers @jywarren I have rebased but due to the preview images fix, the console logs are creating a lot of lag. Even when the page is first loaded, the console logs create a lot of lag. On an old PC/browser the browser might even crash. We will either have to add an option for verbose/non-verbose mode or process everything on an alternate thread. Thoughts?

![console-lag](https://user-images.githubusercontent.com/34770591/51729834-4a4b9d80-209b-11e9-9804-872afc4dc48e.gif)

HarshKhandeparkar commented:
@jywarren can you please merge my PRs before the other ones since I will not be able to touch my PC until 16th of april. I don't know if I will get time past that date as well.

jywarren commented:
Hi! I'm merging this, but I'd like to suggest we put the "close insert step" button in the Insert Step box itself, because where it is now is a bit confusing -- not clear that you're closing the Insert Step box. I think it's OK for now but will open a new issue and hopefully we can get this modified!

Great work @HarshKhandeparkar !!!

...

Hmm, actually now that I am trying to rebase it, I see that some of this was solved in https://github.com/publiclab/image-sequencer/pull/712/, and I am not sure how to proceed here. 

I'm really sorry @HarshKhandeparkar but I think I need to ask you and @Mridul97 to try to figure this out before we can merge here. I can't quite tell what is the "close box" button code and what is the "fixing insert step" part. 🙁  Sorry!!!

HarshKhandeparkar commented:
Jywarren my PR changes the intermediateHtmlStepUi.js file whereas #712 fixes defaultHtmlStepUi.js where the other file was not required properly.

HarshKhandeparkar commented:
@jywarren I just merged with main but it is still showing conflicts.

HarshKhandeparkar commented:
@jywarren I have made some changes but I couldn't check those out. You or someone else will have to check them out locally. Sorry.

jywarren commented:
@Harshithpabbati are you able to test this out perhaps? Or @Divy123 ? Thank you, much appreciated!

Harshithpabbati commented:
> @Harshithpabbati are you able to test this out perhaps? Or @Divy123 ? Thank you, much appreciated!

Just a min i will check this. @jywarren

Divy123 commented:
Checking it .

Harshithpabbati commented:
![box](https://user-images.githubusercontent.com/43822585/54849651-ee07a180-4d0a-11e9-8180-fd1818c253cd.png)
Why are there 3 boxes to close. @HarshKhandeparkar

Harshithpabbati commented:
Sorry that was by mistake. I closed the pull req

Harshithpabbati commented:
just a small change everything is good!!.

Harshithpabbati commented:
But the thing was I can't find that change in pr.

Divy123 commented:
![harsh](https://user-images.githubusercontent.com/31225007/54850290-c74a6a80-4d0c-11e9-8656-62e7441f77b4.png)
Two close buttons appearing nearby each other.
The Add steps button is not appearing and  working.
Please look into these two.
Thanks!!

Harshithpabbati commented:
> ![harsh](https://user-images.githubusercontent.com/31225007/54850290-c74a6a80-4d0c-11e9-8656-62e7441f77b4.png)
> Two close buttons appearing nearby each other.
> The Add steps button is not appearing and working.
> Please look into these two.
> Thanks!!

Add step button is working @Divy123

Harshithpabbati commented:
And other one after adding from add step the button is not changing to add button.
see this 
![add](https://user-images.githubusercontent.com/43822585/54850691-cebe4380-4d0d-11e9-8586-5804bd4d53b0.png)

Harshithpabbati commented:
> And other one after adding from add step the button is not changing to add button.
> see this
> ![add](https://user-images.githubusercontent.com/43822585/54850691-cebe4380-4d0d-11e9-8586-5804bd4d53b0.png)

After adding from the add button then the button shows the vice-versa but functions in the correct way.

Harshithpabbati commented on dist/image-sequencer-ui.js:
```diff
@@ -762,8 +762,9 @@ var urlHash = require('./urlHash.js'),
-    return '<div class="row insertDiv">\
+    return '<div class="row insertDiv collapse">\
         <div class="col-md-6 col-md-offset-2" style="margin-top:5%">\
+	<button class="btn btn-primary close-insert-box"><i class="fa fa-times" aria-hidden="true"> Close</button>\
```
Remove this line.

Harshithpabbati replied:
There are two close buttons.

HarshKhandeparkar commented:
Will fix #918 and #919 soon. Almost done with it locally.

HarshKhandeparkar commented:
@jywarren I am tired rebasing this. Please try to merge it.

![fix-insert-step](https://user-images.githubusercontent.com/34770591/55409850-662e5c80-5580-11e9-9d0c-66ad4ce5d887.gif)

HarshKhandeparkar commented:
I have also fixed the dropdown icon.

HarshKhandeparkar commented:
@jywarren this has been open for 3 months now. I have rebased it several times. Please try to merge it. I have made all the suggested changes like the close button inside the div. Also I have fixed the dropdown icon which wasn't getting rotated.

HarshKhandeparkar commented:
@publiclab/is-reviewers

HarshKhandeparkar commented:
The add step buttons are broken everywhere. Even on beta.sequencer.publiclab.org. That can be fixed in a separate pr.

Harshithpabbati commented:
> The add step buttons are broken everywhere. Even on beta.sequencer.publiclab.org. That can be fixed in a separate pr.

Broken ???. Do you mean it is not visible???

HarshKhandeparkar commented:
It is not responsive

HarshKhandeparkar commented:
Only on xs screens

HarshKhandeparkar commented:
It works well on small screens. Try to open an issue for that.

Harshithpabbati commented:
ok!! I will check it and if needed I will make it an fto.

HarshKhandeparkar commented:
Please review this as well :)

Harshithpabbati commented:
Yeah trying to check it local. :smile:

Harshithpabbati commented:
@HarshKhandeparkar sorry for this. You need to still make few changes as the quick add isn't working in the insert step.

Harshithpabbati commented on examples/demo.js:
```diff
@@ -197,8 +197,8 @@ window.onload = function() {
-      insertPreview.updatePreviews(reader.result,'addStep');
-      insertPreview.updatePreviews(sequencer.steps[0].imgElement.src,'insertStep');
+      insertPreview.updatePreviews(reader.result,'#addStep');
+      insertPreview.updatePreviews(sequencer.steps[0].imgElement.src,'.insertDiv');
```
why did you make it insertDiv???

HarshKhandeparkar replied:
Because earlier it was an id, now it is a class

HarshKhandeparkar commented:
😭

Ok. Can you provide a screenshot/gif(preferably gif)? Also please explain what is not working exactly. Does selectize work? Also what is the error in the console?

HarshKhandeparkar commented:
Or only the quock btns are not working?

HarshKhandeparkar commented:
Ok I know why the previews are not working but O don't know how to fix those. Only @Divy123 does. Can we do that in a followup?

HarshKhandeparkar commented:
The error is not actually because of my pr but because of selectize

Harshithpabbati commented:
> The error is not actually because of my pr but because of selectize

Selectize is working !!. But only the quick add buttons aren't.

HarshKhandeparkar commented:
Divy has fixed the quock btns for the main step selector recently.

Harshithpabbati commented:
And the most important thing is when I added the step it should again change to insert-step but it isn't changing see the close button in the saturation step.
![append](https://user-images.githubusercontent.com/43822585/55412239-b8717c80-5584-11e9-9e0c-0da2931b8b0c.png)

Harshithpabbati commented:
And one close button is enough I guess.

HarshKhandeparkar commented:
Ok this does need some changes. But I cannot do anything anymore.

Harshithpabbati commented:
@HarshKhandeparkar It is looking nice right??? See this this might be the smallest size in mobile.
![xs](https://user-images.githubusercontent.com/43822585/55415294-67fd1d80-558a-11e9-92a1-20aeff15173b.png)

HarshKhandeparkar commented:
:+1:

Harshithpabbati commented:
> @HarshKhandeparkar It is looking nice right??? See this this might be the smallest size in mobile.
> ![xs](https://user-images.githubusercontent.com/43822585/55415294-67fd1d80-558a-11e9-92a1-20aeff15173b.png)

I didn't made any changes.

jywarren commented:
Hi all - @HarshKhandeparkar don't worry, we will get this merged. @Harshithpabbati can you make a checklist of things that remain to be fixed here, and I can see if I can go through and solve them one by one and then merge this?

Don't be discouraged, you've done a tremendous amount here! We are so grateful! 🙌 we'll solve it!

HarshKhandeparkar commented:
I think only one thing is left to be done.

1. When the step is inserted, the following code is to be run.
```js
toggleDiv();
```

That's it.

HarshKhandeparkar commented:
@jywarren I did the changes myself. Done with this. Whew!
Please try to merge this before the others.

@Harshithpabbati try this out locally. 

### Thank you All

HarshKhandeparkar commented:
I just deleted some old comments and removed old omgs because the page was taking too long to load

HarshKhandeparkar commented:
@publiclab/is-reviewers please review :)

Harshithpabbati commented on examples/lib/intermediateHtmlStepUi.js:
```diff
@@ -4,7 +4,6 @@ var urlHash = require('./urlHash.js'),
     return '<div class="row insertDiv collapse">\
-        <div class="col-md-6 col-md-offset-2" style="margin-top:5%">\
           <section class="panel panel-primary .insert-step">\
             <button class="btn btn-default close-insert-box"><i class="fa fa-times" aria-hidden="true"></i> Close</button>\
```
@HarshKhandeparkar remove this line. Because you are appending the close option for two times.

Harshithpabbati replied:
Sorry that's not the line. Remove this line
<button class="btn btn-default close-insert-box"><i class="fa fa-times" aria-hidden="true"></i> Close</button>\

HarshKhandeparkar replied:
Jywarren asked me to do that

Harshithpabbati replied:
ok remove the above line which I mentioned.

HarshKhandeparkar replied:
Jywarren asked me to add a close button inside the box itself.

Harshithpabbati replied:
Oh but it looks odd if we have two buttons for the same option.

HarshKhandeparkar replied:
Most of the times, the other btn will be unnoticed. If by mistake someone opens the box, the btn that is used to open the dialog, can be used to close it.

Harshithpabbati commented:
@HarshKhandeparkar just one more change after changing above mentioned one. It's about the quick add buttons ! They are not working.

Harshithpabbati commented:
@jywarren I have an idea of removing the quick add buttons in the insert-step we can just have the selectize one in the insert step. What are your thoughts about it ?

HarshKhandeparkar commented:
@Divy123 will have to do that. I don't know how to fix it.

HarshKhandeparkar commented:
Divy had fixed it for the main selector. It is some problem with selectize.

HarshKhandeparkar commented:
@jywarren what about this? Will I have to rebase it again because of grid overlay?

Harshithpabbati commented:
@HarshKhandeparkar no need there is no conflict right???

HarshKhandeparkar commented:
I was asking if I'll have to rebase. @jywarren please try to review and merge this.

Harshithpabbati commented:
@jywarren this is ready to merge. And @Divy123 will fix the quick add buttons.

Divy123 commented:
Surely I will.

Harshithpabbati commented:
@jywarren do you think we need two close buttons ?

HarshKhandeparkar commented:
Thanks a lot @Divy123

Harshithpabbati commented:
@jywarren this is ready to merge :smile:

Harshithpabbati commented:
A great work by @harshkhandeparkar :+1:

HarshKhandeparkar commented:
You helped me a lot in it. :heart:
@Harshithpabbati @Divy123 @jywarren

HarshKhandeparkar commented:
This wouldn't have been possibke without the help and feedback everyone provided.

Harshithpabbati commented:
That's just because of open source  💓

jywarren commented:
SOOOOO GREAAAATTTTT!!!! 🎉 🎉 🎉 🎉 🎉 🎉
